### PR TITLE
Qwen2 5 vl vision transformer merging

### PR DESCRIPTION
**Pull request by malkomes**
FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ <https://docs.vllm.ai/en/latest/contributing/overview.html>**

### CODE REVIEW

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 1]
```diff
@@ -72,6 +72,63 @@
...
+    maxsize = indices[-1]
+    range_to_max_for_each_img = torch.arange(
+        maxsize,
+        device=indices.device).unsqueeze(0).repeat(indices.shape[0] - 1, 1)
+    lesser = range_to_max_for_each_img < indices[1:].unsqueeze(1)
+    greater_eq = range_to_max_for_each_img >= indices[:-1].unsqueeze(1)
+    range_indices = torch.logical_and(lesser, greater_eq).float()
+    # can reduce sum externally or as batchmatmul
+    if range_indices.shape[-1] > 40000:
+        logger.info(f"einsum running on CPU : {range_indices.shape}")
```
Comment (sayantan-nervana) [inline vllm/model_executor/models/qwen2_5_vl.py:117]: f-strings evaluated even if logging level is not enabled, so need rewrite as `"einsum running on CPU :" + str(range_indices.shape)`

**Comment by malkomes** [comment]
@michalkuligowski this should be ready to review. Let me know your thoughts

**Comment by imangohari1** [comment]
couple of notes: 
- we have conducted some tests, including online/offline/pytests and to the best of our acknowledge the current tree at this point is funcitonal and performs reasonably with a good accuracy. 
- the documentation for this model is given here: https://github.com/HabanaAI/Gaudi-tutorials/blob/qwen-2.5-vl/PyTorch/vLLM_Tutorials/Benchmarking_on_vLLM/Online_Static/README_qwen-2.5-vl.md

**Comment by malkomes** [comment]
@michalkuligowski can you run the tests again?

**Review by michalkuligowski — APPROVED** [review]

File: requirements-hpu-qwen2_5_vl.txt [hunk 2]
```diff
@@ -1 +1 @@
-transformers @ git+https://github.com/huggingface/transformers.git@6b550462139655d488d4c663086a63e98713c6b9
+transformers @ git+https://github.com/malkomes/transformers.git@ac372cd18f836c41f57cdce46094db00019d4280
```
Comment (michalkuligowski) [inline requirements-hpu-qwen2_5_vl.txt:1]: I dont think we should install custom branches, why ist not from oficial repo?

Comment (malkomes) [reply requirements-hpu-qwen2_5_vl.txt:1]: Yup, this is a custom solution for the 112 alignment assumption that leads to several optimizations that we made for this model. Our goal for the next PR is to remove this requirement. Let me know if this is a blocker for this PR and we can work on anticipating some of these changes.

Comment (sayantan-nervana) [reply requirements-hpu-qwen2_5_vl.txt:1]: This is a requirement only for this model. We need a certain image preprocessing function from transformer (to get 112 alignment). We had tried monkey-patching transformer and ran into some issues. so our goal is to merge this and in the next PR get rid of the 112 alignment requirement or get rid of this dependency (if we have to keep the 112 alignment requirement)

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 3]
```diff
@@ -768,6 +884,185 @@ def load_weights(self, weights: Iterable[Tuple[str,
...
+                )
+
+            pixel_values_curr_img_padded, rot_pos_emb, \
+                cu_seqlens, _, window_index = self.pre_attn(
+                pixel_values_curr_img_padded, img_shape_padded)
+
+            # either a single image,
+            # or a single image and its accompanying pad image,
+            # so only max expansion to 3
+            expanded_cu_seqlens = expand_to_max(cu_seqlens, 3)
```
Comment (sayantan-nervana) [inline vllm/model_executor/models/qwen2_5_vl.py:1042]: @jiminha rightly points out this can be removed now.

Earlier we were passing it in to the fwd pass and the attn mask was created inside. now we create it in line 1047

Earlier we wanted shape of cu_seqlens to be same regardless of padding or not padding. if no padding its len was 2 (0 and imagesize), and with pad it was len = (0, imglen, padding), so we expanded it to sizze 3 to prevent recompilations because it was entering forward (where we wanna prevent recompilations)

In later optimizations we create the mask fullatt_block_attn_mask externally (which is faster to do it only once and needs less memory). so we can remove this line and use cu_seqlens[-1] in the assert in line 1048

**Review by sayantan-nervana — COMMENTED** [review]
@malkomes .. there is the following in the forward of Qwen2_5_VisionTransformer
We could remove this because we have a separate class now?

```
 if is_hpu:
            # NOTE: unique_consecutive is a dynamic operation
            # we are using `remove_duplicates_cpu` instead
            def remove_duplicates_cpu(a):
                return [
                    a[i] for i in range(len(a)) if i == 0 or a[i - 1] != a[i]
                ]

```

**Review by sayantan-nervana — APPROVED** [review]

File: tests/models/registry.py [hunk 4]
```diff
@@ -366,7 +366,8 @@ def check_available_online(
                                                       hf_overrides={"architectures": ["QwenVLForConditionalGeneration"]}),  # noqa: E501
     "Qwen2AudioForConditionalGeneration": _HfExamplesInfo("Qwen/Qwen2-Audio-7B-Instruct"),  # noqa: E501
     "Qwen2VLForConditionalGeneration": _HfExamplesInfo("Qwen/Qwen2-VL-2B-Instruct"),  # noqa: E501
-    "Qwen2_5_VLForConditionalGeneration": _HfExamplesInfo("Qwen/Qwen2.5-VL-3B-Instruct"),  # noqa: E501
+    "Qwen2_5_VLForConditionalGeneration": _HfExamplesInfo("Qwen/Qwen2.5-VL-3B-Instruct",  # noqa: E501
+                                                          min_transformers_version="4.48.9"),  # noqa: E501
```
Comment (madamczyk-intel) [inline tests/models/registry.py:370]: Why is this change needed? It's not specified in upstream code:  https://github.com/vllm-project/vllm/blob/a8f5aec20ad685851f972847c0567db270d9845f/tests/models/registry.py#L379

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 5]
```diff
@@ -326,6 +386,58 @@ def forward(
             context_layer = rearrange(output,
                                       "(b s) ... -> b s ...",
                                       b=batch_size)
+        elif self.attn_backend == _Backend.TORCH_SDPA and is_hpu:
```
Comment (madamczyk-intel) [inline vllm/model_executor/models/qwen2_5_vl.py:389]: Why aren't we using our HpuAttnBackend?

File: vllm/model_executor/models/qwen2_5_vl.py
Comment (madamczyk-intel) [inline vllm/model_executor/models/qwen2_5_vl.py:1]: There's a lot of HPU specific changes here that might be problematic for future upstreaming/maintaining. Please reduce model file changes as much as possible.

**Review by madamczyk-intel — CHANGES_REQUESTED** [review]

File: requirements-hpu-qwen2_5_vl.txt (hunk 2 above)
Comment (madamczyk-intel) [reply requirements-hpu-qwen2_5_vl.txt]: We shouldn't reference custom forks in our code base. For me this is a blocker. @kzawora-intel wdyt?

File: vllm/worker/hpu_model_runner.py [hunk 6]
```diff
@@ -278,6 +308,26 @@ def __init__(self, model, vllm_config, layer_names, is_causal, sampler):
...
+        # the static and dynamic parts distinct.
+
+        if not htorch.utils.internal.is_lazy() and self.model_is_mrope:
+            logger.warning("[Multimodal] HPU is not in Lazy Mode, "
+                           "split graph has not impact")
+
+        if htorch.utils.internal.is_lazy() and self.model_is_mrope:
+            logger.info("[Multimodal] Split Graph to Visual and Language")
+            self.model.visual = htorch.hpu.wrap_in_hpu_graph(
+                self.model.visual, disable_tensor_cache=True)
```
[the remark this answers is not in the document]
Comment (malkomes) [reply vllm/worker/hpu_model_runner.py:327]: yeah, most of the changes are only with `mrope` models.

File: vllm/worker/hpu_model_runner.py [hunk 7]
```diff
@@ -454,17 +504,66 @@ def forward(self, *args, **kwargs):
...
+                image_input = self.model._parse_and_validate_image_input(
+                    **kwargs)
+                video_input = self.model._parse_and_validate_video_input(
+                    **kwargs)
+
+                inputs_embeds = self.model.get_input_embeddings_v0(
+                    input_ids,
+                    image_input=image_input,
+                    video_input=video_input)
+                input_ids = None
```
[the remark this answers is not in the document]
Comment (malkomes) [reply vllm/worker/hpu_model_runner.py:538]: great question! 
- one is that the embedding computation is very dynamic for this model, so this decoupling make it easier to split the preprocessing part from the actual llm run (which we want to wrap it on the hpuGraph). 
- the other is that this is aligned with the v1 changes. We should eventually change the `get_input_embeddings_v0` to `get_input_embeddings`, but this logic of computing the embedding in the model runner is implemented on v1. The reason why we didn't focus too much on making `get_input_embeddings` work is that it requires a scheduler and other changes to the `hpu_model_runner` that are not specific to this model.

File: vllm/model_executor/models/qwen2_5_vl.py
Comment (malkomes) [reply vllm/model_executor/models/qwen2_5_vl.py:1]: Indeed, but most of the changes are restricted to one class `Qwen2_5_VisionTransformerStaticShape` which makes the model more static shape friendly, so that shouldn't be a problem. 

Perhaps, we could move `AttentionLongSequence` to a different location, but I don't see a better place yet.

File: vllm/model_executor/models/qwen2_5_vl.py (hunk 5 above)
Comment (sayantan-nervana) [reply vllm/model_executor/models/qwen2_5_vl.py:389]: this is a copy of the " self.attn_backend == _Backend.TORCH_SDPA " branch (see the next elif branch).

This is for the vision model, and it is done using just regular attention. it doesnnt have any kv cache etc, and image is processed only during prompt phase

File: vllm/worker/hpu_model_runner.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (xuechendi) [reply vllm/worker/hpu_model_runner.py:538]: I see, thanks, how about moving everything inside "if self.model_is_mrope:" to method of HPUModelAdapter then call the method in forward to keep it clean.
Another question is that, the reason for wrapping_as_hpu_graph separately to vision and text model, is that also because you want to avoid dynamic shape capture in this forward function? If so, can we move the get_input_embedding to model_execute who calls model.forward?

[the remark this answers is not in the document]
Comment (malkomes) [reply vllm/worker/hpu_model_runner.py]: @xuechendi your suggestion to move `get_input_embedding` to `model_execute` was fantastic! we were able to simplify several parts of the code with that insight. Thanks!

File: vllm/model_executor/models/qwen2_5_vl.py
Comment (madamczyk-intel) [reply vllm/model_executor/models/qwen2_5_vl.py:1]: Ok, let's keep it as is. If we figure out a better place for it we can move it in the future.

File: requirements-hpu-qwen2_5_vl.txt (hunk 2 above)
Comment (madamczyk-intel) [reply requirements-hpu-qwen2_5_vl.txt]: awesome. Thanks :)

File: vllm/worker/hpu_model_runner.py [hunk 8]
```diff
@@ -1131,6 +1217,15 @@ def move_to_device(self, tensor):
...
 
+    '''
+    Qwen2.5VL requires the bucket's information
+    '''
+
+    def add_vision_buckets_to_model(self):
+        model = self.get_model()
+        if supports_multimodal(model):
+            model.vision_buckets = VisionBuckets()
+
```
Comment (jkaniecki) [inline vllm/worker/hpu_model_runner.py:1228]: If only Qwen require that, why putting it under "if supports_multimodal" ? We shoul keep in mind that there are many multimodal architectures like llama4 or llama3.2. vision (encoder/decoder).

**Review by jkaniecki — COMMENTED** [review]
In my opinion there are to many "multimodal" changes made to basic model runner . MOreover, some of those changes seems to look like Qwen specific, while putting under "general "supports_multimodal" if. I would suggest to make separate model runner / adapter for multimodal models (and another separate for mrope models if this will look more clear) - same is done for enc/dec models (we don't have to keep all in separate files, but it would be good to have separate classes)

File: vllm/worker/hpu_model_runner.py (hunk 8 above)
Comment (malkomes) [reply vllm/worker/hpu_model_runner.py:1228]: If I recall correctly, these models run with `HPUEncoderDecoderModelRunner` and the `create_dummy_seq_group_metadata` in that class creates data for profiling with `multimodal` information. The main difference is that in that case, all images are all resized to be `336x336` vide https://huggingface.co/meta-llama/Llama-4-Scout-17B-16E-Instruct/blob/main/config.json#L61. With a fixed image size resolution we don't need to warmup different images sizes or pad incoming images to predefined buckets. 

One thing I can do is to create a UT similar to this one for `Qwen2.5-vl`. What do you think about that @xuechendi ?

@jkaniecki good point. I will make some changes to make it more specific for `mrope` models. But creating a whole class just for a group of models, rather than an architecture type could be too much. At least the logging information and the warmup for multi-modal scheme can be a template for any other model that accepts images with multiple resolutions. What do you think? I will try to get these changes done tonight.

**Comment by xuechendi** [comment]
@malkomes , this PR failed llama4 multiModel test, please check
image

**Review by sayantan-nervana — APPROVED** [review]
